### PR TITLE
feat(pairing): disable pairing for 2FA accounts

### LINKED ISSUE #6987: Blocking pairing for 2FA users
There was discussion to do it, but then not do it, and then do it again. 

What did we decide? @shane-tomlinson or someone, could you let me know?

### PR DESCRIPTION
**Pull request by vladikoff**
Fixes #6987

### CODE REVIEW

File: app/scripts/lib/auth-errors.js [hunk 1]
```diff
@@ -522,7 +522,11 @@ define(function (require, exports, module) {
     TOKEN_VERIFICATION_CODE_REQUIRED: {
       errno: 1060,
       message: t('Please enter verification code')
-    }
+    },
+    TOTP_PAIRING_NOT_SUPPORTED: {
+      errno: 1061,
+      message: t('Accounts with two-step authentication do not support pairing at this time')
```
Comment (shane-tomlinson) [inline app/scripts/lib/auth-errors.js]: 👍

**Comment by vladikoff** [comment]
@shane-tomlinson should be good to go

File: app/scripts/views/pair/auth_allow.js [hunk 2]
```diff
@@ -16,8 +17,25 @@ class PairAuthAllowView extends FormView {
     'click #cancel': preventDefaultThen('cancel')
   });
 
-  initialize () {
+  beforeRender () {
+    const account = this.getSignedInAccount();
+
     this.listenTo(this.broker, 'error', this.displayError);
+
+    if (! account) {
```
Comment (shane-tomlinson) [inline app/scripts/views/pair/auth_allow.js]: @vladikoff do you want this here on in [pairing/index.js](https://github.com/mozilla/fxa-content-server/blob/2a3120429df99160b3620ae875b0985f524f875a/app/scripts/views/pair/index.js#L28) before the user clicks the button to pair?

**Comment by vladikoff** [comment]
>  I've been trying to add a functional test for this and keep being displayed an "unsupported browser" error

You need to be using latest Nightly and flip the pairing pref on

**Comment by shane-tomlinson** [comment]
> I've been trying to add a functional test for this and keep being displayed an "unsupported browser" error

Mind adding a functional test? I gave myself a short timebox and didn't get it done.

**Comment by shane-tomlinson** [comment]
Here's what I started with:

```diff
commit 86b56c97b07492ddf21ced3080f2d57d76197f3c
Author: Shane Tomlinson <stomlinson@mozilla.com>
Date:   Mon Mar 18 14:18:53 2019 +0000

    .

diff --git a/tests/functional/lib/selectors.js b/tests/functional/lib/selectors.js
index 3abe9aca..e57dad9e 100644
--- a/tests/functional/lib/selectors.js
+++ b/tests/functional/lib/selectors.js
@@ -146,6 +146,7 @@ module.exports = {
   PAIRING: {
     AUTH_SUBMIT: '#auth-approve-btn',
     COMPLETE: '#fxa-404-header',
+    PAIR_FAILURE: '#fxa-pair-failure-header',
     START_PAIRING: '#start-pairing',
     SUPP_SUBMIT: '#supp-approve-btn',
   },
diff --git a/tests/functional/pairing.js b/tests/functional/pairing.js
index 814fbbf1..2172615b 100644
--- a/tests/functional/pairing.js
+++ b/tests/functional/pairing.js
@@ -19,11 +19,13 @@ const BAD_OAUTH_REDIRECT = `${config.fxaOAuthApp}api/oauth`;
 const GOOD_CLIENT_ID = '3c49430b43dfba77';
 const GOOD_PAIR_URL = `${config.fxaContentRoot}pair/supp?response_type=code&client_id=${GOOD_CLIENT_ID}&redirect_uri=${REDIRECT_HOST}oauth%2Fsuccess%2F3c49430b43dfba77&scope=profile%2Bhttps%3A%2F%2Fidentity.mozilla.com%2Fapps%2Foldsync&state=foo&code_challenge_method=S256&code_challenge=IpOAcntLUmKITcxI_rDqMvFTeC9n_g0B8_Pj2yWZp7w&access_type=offline&keys_jwk=eyJjcnYiOiJQLTI1NiIsImt0eSI6IkVDIiwieCI6ImlmcWY2U1pwMlM0ZjA5c3VhS093dmNsbWJxUm8zZXdGY0pvRURpYnc4MTQiLCJ5IjoiSE9LTXh5c1FseExqRGttUjZZbFpaY1Y4MFZBdk9nSWo1ZHRVaWJmYy1qTSJ9`; //eslint-disable-line  max-len
 const BAD_PAIR_URL = `${config.fxaContentRoot}pair/supp?response_type=code&client_id=${BAD_CLIENT_ID}&redirect_uri=${BAD_OAUTH_REDIRECT}&scope=profile%2Bhttps%3A%2F%2Fidentity.mozilla.com%2Fapps%2Foldsync&state=foo&code_challenge_method=S256&code_challenge=IpOAcntLUmKITcxI_rDqMvFTeC9n_g0B8_Pj2yWZp7w&access_type=offline&keys_jwk=eyJjcnYiOiJQLTI1NiIsImt0eSI6IkVDIiwieCI6ImlmcWY2U1pwMlM0ZjA5c3VhS093dmNsbWJxUm8zZXdGY0pvRURpYnc4MTQiLCJ5IjoiSE9LTXh5c1FseExqRGttUjZZbFpaY1Y4MFZBdk9nSWo1ZHRVaWJmYy1qTSJ9`; //eslint-disable-line  max-len
+const SETTINGS_URL = `${config.fxaContentRoot}settings`;
 
 const PASSWORD = 'PASSWORD123123';
 let email;
 
 const {
+  confirmTotpCode,
   createUser,
   click,
   closeCurrentWindow,
@@ -140,6 +142,41 @@ registerSuite('pairing', {
         .then(closeCurrentWindow());
     },
 
+    'does not allow 2FA enabled accounts to pair': function () {
+      email = TestHelpers.createEmail();
+
+      return this.remote
+        .then(createUser(email, PASSWORD, { preVerified: true }))
+        .then(openPage(SIGNIN_PAGE_URL, selectors.ENTER_EMAIL.HEADER))
+        .then(type(selectors.ENTER_EMAIL.EMAIL, email))
+        .then(click(selectors.ENTER_EMAIL.SUBMIT, selectors.SIGNIN_PASSWORD.HEADER))
+        .then(type(selectors.SIGNIN_PASSWORD.PASSWORD, PASSWORD))
+        .then(click(selectors.SIGNIN_PASSWORD.SUBMIT, selectors.CONNECT_ANOTHER_DEVICE.HEADER))
+        // but the login message is sent automatically.
+        .then(testIsBrowserNotified('fxaccounts:login'))
+
+        .then(openPage(SETTINGS_URL, selectors.TOTP.MENU_BUTTON))
+
+        .then(click(selectors.TOTP.MENU_BUTTON))
+        .then(click(selectors.TOTP.SHOW_CODE_LINK))
+
+        .then(testElementExists(selectors.TOTP.QR_CODE))
+
+        .then(testElementExists(selectors.TOTP.MANUAL_CODE))
+
+        // Store the secret key to recalculate the code later
+        .findByCssSelector(selectors.TOTP.MANUAL_CODE)
+        .getVisibleText()
+        .then(confirmTotpCode)
+        .end()
+
+        .then(openPage(`${config.fxaContentRoot}pair`, selectors.PAIRING.START_PAIRING))
+        .then(click(selectors.PAIRING.START_PAIRING))
+
+        .then(waitForQR())
+        .then(testElementExists(selectors.PAIRING.PAIR_FAILURE));
+    },
+
     'handles invalid clients': function () {
       return this.remote
         .then(openPage(`${BAD_PAIR_URL}#channel_id=foo&channel_key=bar`, selectors['400'].ERROR))
```

File: app/scripts/views/pair/auth_allow.js (hunk 2 above)
Comment (vladikoff) [reply app/scripts/views/pair/auth_allow.js]: We should discuss this, we can block it here or on /pair. If it is on /pair then there is a sneaky way you can still pair with 2FA (by going to a specific about:prefs url)

**Comment by vladikoff** [comment]
@shane-tomlinson the test is mostly fine (i think it's missing 2fa confirmation), but the bigger problem is logging out the browser from the previous account. Any ideas?

I was trying with something like this, but no luck:

```
        .execute(() => {
          const event = new window.CustomEvent('WebChannelMessageToChrome', {
            detail: {
              id: 'account_updates',
              message: {
                command: 'fxaccounts:logout',
                data: {}
              }
            }
          });

          window.dispatchEvent(event);
        })
```

I also cannot logout via `about:preference#sync` because there is a model dialog that we cannot click on with functional tests

**Comment by shane-tomlinson** [comment]
@vladikoff - Nice solution! 

```
     .execute(() => {
          const event = new window.CustomEvent('WebChannelMessageToChrome', {
            detail: {
              id: 'account_updates',
              message: {
                command: 'fxaccounts:logout',
                data: {}
              }
            }
          });

          window.dispatchEvent(event);
        })
```

For the `logout` message to work, you also have to [send a uid](https://searchfox.org/mozilla-central/rev/56705678f5fc363be5e0237e1686f619b0d23009/services/fxaccounts/FxAccountsWebChannel.jsm#155) in the data.

**Comment by vladikoff** [comment]
@shane-tomlinson check now, sadly we cannot do a separate functional test due to merge warnings. Instead I extended the existing test for now.

File: tests/functional/pairing.js [hunk 3]
```diff
@@ -137,7 +139,22 @@ registerSuite('pairing', {
...
+        .then(click(selectors.TOTP.SHOW_CODE_LINK))
+
+        .findByCssSelector(selectors.TOTP.MANUAL_CODE)
+        .getVisibleText()
+        .then((secretKey) => {
+          return this.remote.then(confirmTotpCode(secretKey));
+        })
+        .end()
+
+        .then(openPage(`${config.fxaContentRoot}pair`, selectors.PAIRING.PAIR_FAILURE));
```
Comment (shane-tomlinson) [inline tests/functional/pairing.js]: 👍

**Pull request merged**